### PR TITLE
add support for datastream lifecycle

### LINKED ISSUE #673: [Feature] ES.IndexTemplate.DataStream.Lifecycle.data_retention
**Describe the resource you would like to have implemented.**
DataStream Lifecyle GA'd v8.14.0 and its [bootstrapping](https://www.elastic.co/guide/en/elasticsearch/reference/current/tutorial-manage-new-data-stream.html#create-index-template-with-lifecycle) allows setting `lifecycle.data_retention` which is not explicitly allowed Terraform-side under [Index Template](https://registry.terraform.io/providers/elastic/elasticstack/latest/docs/resources/elasticsearch_index_template). Speculatively, [this code ballpark](https://github.com/elastic/terraform-provider-elasticstack/blob/eb077fd2993321ec813222c317aed0a831c6f073/internal/elasticsearch/index/template.go#L149). 

**Describe alternatives you've considered**
Per Dev internal confirmation, this [alternative doc](https://registry.terraform.io/providers/elastic/elasticstack/latest/docs/resources/elasticsearch_data_stream) suffices as stop-gap as you can actually put it under `settings`.

### PR DESCRIPTION
**Pull request by aizerin**
resolves #673

### CODE REVIEW

File: internal/elasticsearch/index/template.go [hunk 1]
```diff
@@ -298,6 +313,14 @@ func expandTemplate(config interface{}) (models.Template, bool, diag.Diagnostics
 	}
 	templ.Aliases = aliases
 
+	lifecycle, diags := ExpandLifecycle(definedTempl["lifecycle"].(*schema.Set))
+	if diags.HasError() {
+		return templ, false, diags
+	}
+	if lifecycle != nil {
+		templ.Lifecycle = lifecycle
+	}
```
Comment (tobio) [inline internal/elasticsearch/index/template.go]: The tests are failing when `lifecycle` is not defined in the TF module. I'm pretty sure this should fix it, though it's totally untested. 
```suggestion
	if lc, ok := definedTempl["lifecycle"]; ok {
		lifecycle, diags := ExpandLifecycle(lc.(*schema.Set))
		if diags.HasError() {
			return templ, false, diags
		}
		if lifecycle != nil {
			templ.Lifecycle = lifecycle
		}
	}
```

File: internal/elasticsearch/index/commons.go [hunk 2]
```diff
@@ -80,3 +80,30 @@ func FlattenIndexAlias(name string, alias models.IndexAlias) (interface{}, diag.
 
 	return a, diags
 }
+
+func ExpandLifecycle(definedLifecycle *schema.Set) (*models.LifecycleSettings, diag.Diagnostics) {
```
Comment (tobio) [inline internal/elasticsearch/index/commons.go]:
```suggestion
func ExpandLifecycle(definedLifecycle *schema.Set) *models.LifecycleSettings {
```

There's never any actual diags returned here.

File: internal/elasticsearch/index/commons.go [hunk 3]
```diff
@@ -80,3 +80,30 @@ func FlattenIndexAlias(name string, alias models.IndexAlias) (interface{}, diag.
...
+		lifecycle := &models.LifecycleSettings{}
+		if s, ok := lifecycleMap["data_retention"]; ok {
+			lifecycle.DataRetention = s.(string)
+		}
+		return lifecycle, diags
+	}
+	return nil, diags
+}
+
+func FlattenLifecycle(lifecycle *models.LifecycleSettings) (interface{}, diag.Diagnostics) {
```
Comment (tobio) [inline internal/elasticsearch/index/commons.go]:
```suggestion
func FlattenLifecycle(lifecycle *models.LifecycleSettings) interface{} {
```

Similarly, diags is unused.

**Review by tobio — COMMENTED** [review]
Thanks for adding this in. one small change to make this work when `lifecycle` isn't defined.

**Comment by tobio** [comment]
@aizerin are you able to add an entry to the [CHANGELOG](https://github.com/elastic/terraform-provider-elasticstack/blob/main/CHANGELOG.md) for this one as well please?

File: internal/elasticsearch/index/template.go (hunk 1 above)
Comment (aizerin) [reply internal/elasticsearch/index/template.go]: Sorry for mistake. I didn't test newly created index. I fixed it in `ExpandLifecycle` function.

**Comment by aizerin** [comment]
I have pushed one more fix. Tests should work now.

**Review by tobio — APPROVED** [review]
LGTM. Thanks for taking the time to get this added in!

**Pull request merged**